cpu/stm32f1/stmclk.c

```diff
@@ -0,0 +1,209 @@
+#error "Please provide a valid PLL configuration in your board's periph_conf.h"
+#endif
+#if !defined(CLOCK_AHB_DIV) || !defined(CLOCK_AHB) || \
+    !defined(CLOCK_APB1_DIV) || !defined(CLOCK_APB1) || \
+    !defined(CLOCK_APB2_DIV) || !defined(CLOCK_APB2)
+#error "Please provide a AHB and APBx configuration in your board configuration"
+#endif
+
+/* make sure the selected system clock is valid */
+#if (CLOCK_CORECLOCK > 72000000)
+#error "clock config: the selected system clock exceeds 72MHz"
+#endif
+
+/* figure out which base block to use */
+#if CLOCK_HSE
+#if (CLOCK_HSE < 4000000) || (CLOCK_HSE > 16000000)
+#error "clock config: HSE value is out of valid range"
+#endif
+#define BASECLK                 (CLOCK_HSE)
+#else
+#define BASECLK                 (8000000)       /* HSI is alway 8MHz */
+#endif
+
+/* if PLL is configured, verify its parameters */
+#if (CLOCK_PLL_DIV && CLOCK_PLL_MUL)
+#define USEPLL
+/* check clock config */
+#if (CLOCK_CORECLOCK != ((BASECLK / CLOCK_PLL_DIV) * CLOCK_PLL_MUL))
+#error "clock config: PLL configuration does not yield expected system clock"
+#endif
+/* make sure PLL_MUL is in range */
+#if (CLOCK_PLL_MUL < 2) || (CLOCK_PLL_MUL > 17)
+#error "clock config: CLOCK_PLL_MUL is out of range"
+#endif
+/* make sure PLL_DIV is 2 when using HSI as input */
+#if (!CLOCK_HSE) && (CLOCK_PLL_DIV != 2)
+#error "clock config: CLOCK_PLL_DIV must be 2 when using HSI oscillator"
+#endif
+/* and produce the actual PLL configuration */
+#if CLOCK_HSE
+#define PLLSRC                  (RCC_CFGR_PLLSRC)
+#if (CLOCK_PLL_DIV == 2)
+#define PLLDIV                  (RCC_CFGR_PLLXTPRE)
+#else
+#define PLLDIV                  (0)
+#endif
+#else
+#define PLLSRC                  (0)
+#define PLLDIV                  (0)
+#endif
+#define PLLMUL                  ((CLOCK_PLL_MUL - 2) << 18)
+/* and join it for writing the the CFGR register */
+#define PLLCFG                  (PLLMUL | PLLDIV | PLLSRC)
+#else
+#define PLLCFG                  (0)
+#endif
+
+/* now we need to select the system clock source configuration */
+#ifdef USEPLL
+#define SYSCLK_SRC              RCC_CFGR_SW_PLL
+#define SYSCLK_BSY              RCC_CFGR_SWS_PLL
+#elif CLK_HSE
+#define SYSCLK_SRC              RCC_CFGR_SW_HSE
+#define SYSCLK_BSY              RCC_CFGR_SWS_HSE
+#elif
+#define SYSCLK_SRC              RCC_CFGR_SW_HSI
+#define SYSCLK_BSY              RCC_CFGR_SWS_HSI
+#endif
+
+/* Configuration of flash access cycles */
+#define FLASH_WAITSTATES        ((CLOCK_CORECLOCK - 1) / 24000000U)
+
+/* define some bitfields */
+#define HSITRIM                 (1 << 7)
+#define SWSHSI                  (0)
+
+
+void stmclk_init_sysclk(void)
+{
+    /* disable any IRQs */
+    unsigned is = irq_disable();
+    RCC->CIR = 0;
+
+    /* enable HSI and use it as system clock */
+    stmclk_enable_hsi();
+    RCC->CFGR &= ~(RCC_CFGR_SW);
+    while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_HSI) {}
+
+    /* its save now to program the flash wait states */
```

Comment:
spellings seems correct. Meaning: say you had a (fast) PLL configured as clock in the bootloader, but want to switch to some significant slower clock in some image you boot into. Then you would break the code if you would reprogram the flash wait states to some slower value while the PLL is still driving the system. But now that we switched to the HSI as system clock, we can program the flash wait states to whatever we like without breaking anything...